vmdb/app/models/ext_management_system.rb

```diff
@@ -1,20 +1,26 @@
 class ExtManagementSystem < ActiveRecord::Base
+  SUBCLASSES = %w(
+    EmsInfra
+    EmsCloud
+    EmsContainer
+  )
+
   include EmsRefresh::Manager
 
   def self.types
-    [EmsInfra, EmsCloud].collect(&:types).flatten
+    leaf_subclasses.collect(&:ems_type)
   end
 
   def self.supported_types
-    [EmsInfra, EmsCloud].collect(&:supported_types).flatten
+    supported_subclasses.collect(&:ems_type)
   end
 
   def self.leaf_subclasses
-    [EmsInfra, EmsCloud].collect(&:subclasses).flatten
+    descendants.select { |d| d.subclasses.empty? }
   end
 
   def self.supported_subclasses
-    [EmsInfra, EmsCloud].collect(&:supported_subclasses).flatten
+    subclasses.flat_map(&:supported_subclasses)
```

Comment:
Ok, we need code ttl... it should self destruct if it's not deleted by the time we decided it should be deleted ;-)